Repository method `updateRelatedEntity` won't compile

After upgrading from micronaut 2.4.x to 2.5.x a repository method that updates a related entity does not compile anymore.

### Task List

- [x] Steps to reproduce provided
- [x] Stacktrace (if present) provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. Create an app with mn-data-jdbc
2. Create two entities where one relates to the other
    ```groovy
   @MappedEntity
   class EntityOne {

       @Id
       @AutoPopulated
       UUID id

       @NotBlank
       String name

       @NotNull
       EntityTwo entityTwo
   }

   @MappedEntity
   class EntityTwo {

       @Id
       @AutoPopulated
       UUID id

       @NotBlank
       String name
   }
    ```
3. Create a repo with a method that updates the releated entity:
    ```groovy
    @JdbcRepository(dialect = Dialect.H2)
    interface TestRepo extends CrudRepository<EntityOne,UUID> {
        void updateEntityTwo(@Id UUID id, @NotNull @Valid EntityTwo entityTwo)
    }
    ```
4. Run the app

### Expected Behaviour

The app will start without errors and the repostoty method should work.

### Actual Behaviour

The app won't start.

The error is as below:

```
startup failed:
/home/sf/src/issues/mn-data-extended-repo/src/main/groovy/com/example/TestRepo.groovy: 13: Unable to implement Repository method: TestRepo.updateEntityTwo(UUID id,EntityTwo entityTwo). Cannot find a parameter representing the entity update
 @ line 13, column 5.
       void updateEntityTwo(@Id UUID id, @NotNull @Valid EntityTwo entityTwo)
       ^

1 error
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':compileGroovy'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:187)
        at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:268)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:185)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:173)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:109)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:76)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:76)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:74)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:408)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:395)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:388)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:374)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: org.gradle.api.internal.tasks.compile.CompilationFailedException: Compilation failed; see the compiler error output for details.
        at org.gradle.api.internal.tasks.compile.ApiGroovyCompiler.execute(ApiGroovyCompiler.java:283)
        at org.gradle.api.internal.tasks.compile.ApiGroovyCompiler.execute(ApiGroovyCompiler.java:67)
        at org.gradle.api.internal.tasks.compile.GroovyCompilerFactory$DaemonSideCompiler.execute(GroovyCompilerFactory.java:98)
        at org.gradle.api.internal.tasks.compile.GroovyCompilerFactory$DaemonSideCompiler.execute(GroovyCompilerFactory.java:77)
        at org.gradle.api.internal.tasks.compile.daemon.AbstractDaemonCompiler$CompilerWorkAction.execute(AbstractDaemonCompiler.java:135)
        at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
        at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:49)
        at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:43)
        at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:97)
        at org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:43)
        at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:49)
        at org.gradle.workers.internal.IsolatedClassloaderWorker.run(IsolatedClassloaderWorker.java:30)
        at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:85)
        at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:55)
        at org.gradle.process.internal.worker.request.WorkerAction$1.call(WorkerAction.java:138)
        at org.gradle.process.internal.worker.child.WorkerLogEventListener.withWorkerLoggingProtocol(WorkerLogEventListener.java:41)
        at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:135)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
        at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
```

### Environment Information

- **Operating System**: Linux
- **Micronaut Version:** 2.5.4
- **JDK Version:** 11

### Example Application

The README provides info how to reproduce and a branch with micronaut 2.4.4 that shows that it was working before.

- https://github.com/safri-net/mn-2.5-update-relasted-entity-issue

--- a/data-hibernate-jpa/src/test/groovy/io/micronaut/data/hibernate/AbstractHibernateQuerySpec.groovy
+++ b/data-hibernate-jpa/src/test/groovy/io/micronaut/data/hibernate/AbstractHibernateQuerySpec.groovy
@@ -409,12 +409,24 @@ abstract class AbstractHibernateQuerySpec extends AbstractQuerySpec {
             author.books.size() == 1
     }
 
+    void "test update relation custom query"() {
+        when:
+            def book = bookRepository.findAllByTitleStartsWith("Along Came a Spider").first()
+            def author = authorRepository.searchByName("Stephen King")
+            bookRepository.updateAuthorCustomQuery(book.id, author)
+            book = bookRepository.findById(book.id).get()
+        then:
+            book.author.id == book.author.id
+    }
+
     void "test update relation"() {
         when:
             def book = bookRepository.findAllByTitleStartsWith("Along Came a Spider").first()
-            bookRepository.updateAuthor(book.id, book.author)
+            def author = authorRepository.searchByName("Stephen King")
+            bookRepository.updateAuthor(book.id, author)
+            book = bookRepository.findById(book.id).get()
         then:
-            book.publisher == book.publisher
+            book.author.id == book.author.id
     }
 
     void "test query by relation"() {

--- a/data-hibernate-jpa/src/test/java/io/micronaut/data/hibernate/BookRepository.java
+++ b/data-hibernate-jpa/src/test/java/io/micronaut/data/hibernate/BookRepository.java
@@ -15,6 +15,7 @@
  */
 package io.micronaut.data.hibernate;
 
+import io.micronaut.data.annotation.Id;
 import io.micronaut.data.annotation.Query;
 import io.micronaut.data.annotation.Repository;
 import io.micronaut.data.annotation.Where;
@@ -67,7 +68,9 @@ public abstract class BookRepository extends io.micronaut.data.tck.repositories.
     public abstract long updateCustomOnlyTitles(Collection<Book> books);
 
     @Query("UPDATE Book SET author = :author WHERE id = :id")
-    public abstract long updateAuthor(Long id, Author author);
+    public abstract long updateAuthorCustomQuery(Long id, Author author);
+
+    public abstract long updateAuthor(@Id Long id, Author author);
 
     @Query("SELECT b FROM Book b WHERE b.author = :author")
     public abstract List<Book> findByAuthor(Author author);

--- a/data-processor/src/main/java/io/micronaut/data/processor/visitors/RepositoryTypeElementVisitor.java
+++ b/data-processor/src/main/java/io/micronaut/data/processor/visitors/RepositoryTypeElementVisitor.java
@@ -552,7 +552,10 @@ public class RepositoryTypeElementVisitor implements TypeElementVisitor<Reposito
 
     private DataType[] getDataTypes(MethodMatchContext methodMatchContext, ParameterElement[] parameters, List<QueryParameterBinding> finalParameterBinding) {
         Map<String, DataType> parametersTypes = Arrays.stream(parameters)
-                .map(p -> new AbstractMap.SimpleEntry<>(p.getName(), TypeUtils.resolveDataType(p).orElse(null)))
+                .map(p -> new AbstractMap.SimpleEntry<>(
+                        p.stringValue(Parameter.class).orElse(p.getName()),
+                        TypeUtils.resolveDataType(p).orElse(null))
+                )
                 .filter(e -> e.getValue() != null)
                 .collect(Collectors.toMap(AbstractMap.SimpleEntry::getKey, AbstractMap.SimpleEntry::getValue));
 
@@ -697,7 +700,8 @@ public class RepositoryTypeElementVisitor implements TypeElementVisitor<Reposito
     }
 
     private DataType resolvePropertyDataType(MethodMatchContext matchContext, ParameterElement[] parameters, String value) {
-        Map<String, ParameterElement> paramMap = Arrays.stream(parameters).collect(Collectors.toMap(Element::getName, p -> p));
+        Map<String, ParameterElement> paramMap = Arrays.stream(parameters)
+                .collect(Collectors.toMap(e -> e.stringValue(Parameter.class).orElse(e.getName()), p -> p));
         int dot = value.indexOf('.');
         if (dot > -1) {
             String val = value.substring(0, dot);

--- a/data-processor/src/main/java/io/micronaut/data/processor/visitors/finders/RawQueryMethod.java
+++ b/data-processor/src/main/java/io/micronaut/data/processor/visitors/finders/RawQueryMethod.java
@@ -15,6 +15,7 @@
  */
 package io.micronaut.data.processor.visitors.finders;
 
+import io.micronaut.context.annotation.Parameter;
 import io.micronaut.core.annotation.Introspected;
 import io.micronaut.core.annotation.NonNull;
 import io.micronaut.core.annotation.Nullable;
@@ -214,7 +215,9 @@ public class RawQueryMethod implements MethodCandidate {
         if (namedParameters) {
             while (matcher.find()) {
                 String name = matcher.group(3);
-                Optional<ParameterElement> element = parameters.stream().filter(p -> p.getName().equals(name)).findFirst();
+                Optional<ParameterElement> element = parameters.stream()
+                        .filter(p -> p.stringValue(Parameter.class).orElse(p.getName()).equals(name))
+                        .findFirst();
                 if (element.isPresent()) {
                     parameterBindings.add(QueryParameterBinding.of(name, name, DataType.OBJECT, new QueryParameter(name)));
                 } else if (persistentEntity != null) {
@@ -225,7 +228,7 @@ public class RawQueryMethod implements MethodCandidate {
                         parameterBindings.add(QueryParameterBinding.of(name, propertyPath.getPath(), propertyPath.getProperty().getDataType()));
                     }
                 } else {
-                    matchContext.fail("No method parameter found for named Query parameter : " + name);
+                    matchContext.fail("No method parameter found for named Query parameter: " + name);
                     return null;
                 }
             }
@@ -233,7 +236,9 @@ public class RawQueryMethod implements MethodCandidate {
             int index = 1;
             while (matcher.find()) {
                 String name = matcher.group(3);
-                Optional<ParameterElement> element = parameters.stream().filter(p -> p.getName().equals(name)).findFirst();
+                Optional<ParameterElement> element = parameters.stream()
+                        .filter(p -> p.stringValue(Parameter.class).orElse(p.getName()).equals(name))
+                        .findFirst();
                 if (element.isPresent()) {
                     parameterBindings.add(QueryParameterBinding.of(String.valueOf(index++), name, DataType.OBJECT, new QueryParameter(name)));
                 } else if (persistentEntity != null) {
@@ -244,7 +249,7 @@ public class RawQueryMethod implements MethodCandidate {
                         parameterBindings.add(QueryParameterBinding.of(String.valueOf(index++), propertyPath.getPath(), propertyPath.getProperty().getDataType()));
                     }
                 } else {
-                    matchContext.fail("No method parameter found for named Query parameter : " + name);
+                    matchContext.fail("No method parameter found for named Query parameter: " + name);
                     return null;
                 }
             }

--- a/data-processor/src/main/java/io/micronaut/data/processor/visitors/finders/UpdateMethod.java
+++ b/data-processor/src/main/java/io/micronaut/data/processor/visitors/finders/UpdateMethod.java
@@ -61,6 +61,11 @@ public class UpdateMethod extends AbstractPatternBasedMethod {
     }
 
     @Override
+    public int getOrder() {
+        return DEFAULT_POSITION - 150;
+    }
+
+    @Override
     public boolean isMethodMatch(MethodElement methodElement, MatchContext matchContext) {
         boolean isMatch = super.isMethodMatch(methodElement, matchContext) &&
                 methodElement.getParameters().length > 1 &&
@@ -113,17 +118,17 @@ public class UpdateMethod extends AbstractPatternBasedMethod {
         ParameterElement versionMatchMethodParameter = null;
 
         QueryModel query = QueryModel.from(entity);
-        query.idEq(new QueryParameter(idParameter.getName()));
+        query.idEq(new QueryParameter(getParameterName(idParameter)));
         SourcePersistentProperty version = entity.getVersion();
         if (version != null) {
             versionMatchMethodParameter = parameters.stream().filter(p -> p.hasAnnotation(Version.class)).findFirst().orElse(null);
             if (versionMatchMethodParameter != null) {
-                query.versionEq(new QueryParameter(versionMatchMethodParameter.getName()));
+                query.versionEq(new QueryParameter(getParameterName(versionMatchMethodParameter)));
             }
         }
         List<String> propertiesToUpdate = new ArrayList<>(remainingParameters.size());
         for (ParameterElement parameter : remainingParameters) {
-            String name = parameter.stringValue(Parameter.class).orElse(parameter.getName());
+            String name = getParameterName(parameter);
             SourcePersistentProperty prop = entity.getPropertyByName(name);
             if (prop == null) {
                 matchContext.fail("Cannot update non-existent property: " + name);
@@ -157,14 +162,16 @@ public class UpdateMethod extends AbstractPatternBasedMethod {
                 propertiesToUpdate.stream().toArray(String[]::new)
         );
 
-        info.addParameterRole(TypeRole.ID,
-                idParameter.getName()
-        );
+        info.addParameterRole(TypeRole.ID, getParameterName(idParameter));
 
         if (versionMatchMethodParameter != null) {
             info.setOptimisticLock(true);
         }
         return info;
+    }
+
+    private String getParameterName(ParameterElement p) {
+        return p.stringValue(Parameter.class).orElse(p.getName());
     }
 
 }

--- a/data-processor/src/test/groovy/io/micronaut/data/processor/sql/BuildUpdateSpec.groovy
+++ b/data-processor/src/test/groovy/io/micronaut/data/processor/sql/BuildUpdateSpec.groovy
@@ -255,4 +255,40 @@ interface CompanyRepository extends CrudRepository<Company, Long> {
             updateByLastUpdatedMethod.stringValues(DataMethod, DataMethod.META_MEMBER_PARAMETER_BINDING) == ["-1", "0"]
             updateByLastUpdatedMethod.enumValues(DataMethod, DataMethod.META_MEMBER_PARAMETER_TYPE_DEFS, DataType) == [DataType.TIMESTAMP, DataType.TIMESTAMP]
     }
+
+    void "test build update relation"() {
+        given:
+            def repository = buildRepository('test.BookRepository', """
+import io.micronaut.data.jdbc.annotation.JdbcRepository;
+import io.micronaut.data.model.query.builder.sql.Dialect;
+import io.micronaut.data.tck.entities.Book;
+import io.micronaut.data.tck.entities.Author;
+
+@JdbcRepository(dialect= Dialect.MYSQL)
+@io.micronaut.context.annotation.Executable
+interface BookRepository extends CrudRepository<Book, Long> {
+
+    @Query("UPDATE Book SET author_id = :author WHERE id = :id")
+    long updateAuthorCustom(Long id, Author author);
+
+    long updateAuthor(@Id Long id, Author author);
+
+}
+""")
+        when:
+            def updateAuthorCustomMethod = repository.findPossibleMethods("updateAuthorCustom").findFirst().get()
+        then:
+            updateAuthorCustomMethod.stringValue(Query).get() == 'UPDATE Book SET author_id = :author WHERE id = :id'
+            updateAuthorCustomMethod.stringValue(Query, "rawQuery").get() == 'UPDATE Book SET author_id = ? WHERE id = ?'
+            updateAuthorCustomMethod.stringValues(DataMethod, DataMethod.META_MEMBER_PARAMETER_BINDING) == ["1", "0"] as String[]
+            updateAuthorCustomMethod.enumValues(DataMethod, DataMethod.META_MEMBER_PARAMETER_TYPE_DEFS, DataType) == [DataType.ENTITY, DataType.LONG]
+        when:
+            def updateAuthorMethod = repository.findPossibleMethods("updateAuthor").findFirst().get()
+        then:
+            updateAuthorMethod.stringValue(Query).get() == 'UPDATE `book` SET `author_id`=? WHERE (`id` = ?)'
+            updateAuthorMethod.stringValues(DataMethod, DataMethod.META_MEMBER_PARAMETER_BINDING) == ["-1", "0"] as String[]
+            updateAuthorMethod.stringValues(DataMethod, DataMethod.META_MEMBER_PARAMETER_BINDING_PATHS) == ["1.id", ""] as String[]
+            updateAuthorMethod.enumValues(DataMethod, DataMethod.META_MEMBER_PARAMETER_TYPE_DEFS, DataType) == [DataType.LONG, DataType.LONG]
+
+    }
 }

--- a/data-tck/src/main/groovy/io/micronaut/data/tck/tests/AbstractRepositorySpec.groovy
+++ b/data-tck/src/main/groovy/io/micronaut/data/tck/tests/AbstractRepositorySpec.groovy
@@ -1553,6 +1553,30 @@ abstract class AbstractRepositorySpec extends Specification {
             studentRepository.deleteAll()
     }
 
+    void "test update relation custom query"() {
+        given:
+            setupBooks()
+        when:
+            def book = bookRepository.findAllByTitleStartingWith("Along Came a Spider").first()
+            def author = authorRepository.searchByName("Stephen King")
+            bookRepository.updateAuthorCustom(book.id, author)
+            book = bookRepository.findById(book.id).get()
+        then:
+            book.author.id == book.author.id
+    }
+
+    void "test update relation"() {
+        given:
+            setupBooks()
+        when:
+            def book = bookRepository.findAllByTitleStartingWith("Along Came a Spider").first()
+            def author = authorRepository.searchByName("Stephen King")
+            bookRepository.updateAuthor(book.id, author)
+            book = bookRepository.findById(book.id).get()
+        then:
+            book.author.id == book.author.id
+    }
+
     private GregorianCalendar getYearMonthDay(Date dateCreated) {
         def cal = dateCreated.toCalendar()
         def localDate = LocalDate.of(cal.get(Calendar.YEAR), cal.get(Calendar.MONTH) + 1, cal.get(Calendar.DAY_OF_MONTH))

--- a/data-tck/src/main/java/io/micronaut/data/tck/repositories/BookRepository.java
+++ b/data-tck/src/main/java/io/micronaut/data/tck/repositories/BookRepository.java
@@ -96,6 +96,11 @@ public abstract class BookRepository implements PageableRepository<Book, Long> {
     @Query(value = "select * from book where (CASE WHEN exists ( select (:arg0) ) THEN title = ANY (:arg0) ELSE true END)", nativeQuery = true)
     public abstract List<Book> listNativeBooksNullableArrayAsStringArray(@Nullable @TypeDef(type = DataType.STRING_ARRAY) String[] arg0);
 
+    @Query("UPDATE book SET author_id = :author WHERE id = :id")
+    public abstract long updateAuthorCustom(@Parameter("id") Long id, @Parameter("author") Author author);
+
+    public abstract long updateAuthor(@Parameter("id") @Id Long id, @Parameter("author") Author author);
+
     public void saveAuthorBooks(List<AuthorBooksDto> authorBooksDtos) {
         List<Author> authors = new ArrayList<>();
         for (AuthorBooksDto dto: authorBooksDtos) {

